libbeat/common/mapstr.go

```diff
@@ -277,28 +277,57 @@ func MergeFields(ms, fields MapStr, underRoot bool) error {
 		return nil
 	}
 
-	fieldsMS := ms
+	destMap, err := mergeFieldsGetDestMap(ms, fields, underRoot)
+	if err != nil {
+		return err
+	}
+
+	// Add fields and override.
+	for k, v := range fields {
+		destMap[k] = v
+	}
+
+	return nil
+}
+
+// MergeFieldsDeep recursively merges the keys and values from fields into ms, either
+// into ms itself (if underRoot == true) or into ms["fields"] (if underRoot == false). If
+// the same key exists in fields and the destination map, the value in fields takes precedence.
+//
+// An error is returned if underRoot is true and the value of ms["fields"] is not a
+// MapStr.
+func MergeFieldsDeep(ms, fields MapStr, underRoot bool) error {
```

Comment:
Yay, more methods/functions operating on MapStr :(

Let's rename 'ms' to 'target' and 'fields' to 'from', to make it more clear that we do overwrite fields in the original MapStr.

Note: I read a 'merge' function as something that does not destructively modifies one of its inputs.

Reply:
> Note: I read a 'merge' function as something that does not destructively modifies one of its inputs.

I was keeping the naming and behavior consistent with the older `MergeFields` function, to try and indicate that `MergeFieldsDeep` behaves just like `MergeFields` (it modifies one of the inputs) but in a recursive/deep fashion.

I'm happy to change the name to something else or make the function return a new `MapStr` instead of modifying an input. Or perhaps we should make this a method on `MapStr`? To me, it doesn't seem generic enough to become a method (since it does some very specific stuff with the `fields` field) which is why I stayed away from that option.

Reply:
I'm ok keeping the name `MergeFieldsDeep`.